ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh

```diff
@@ -685,7 +685,23 @@ function upgrade() {
         clear_upgrade
         check_upgrade_status "${cluster_src_ver}"
     fi
-    run_command "oc adm upgrade --to-image=${TARGET} --allow-explicit-upgrade --force=${FORCE_UPDATE}"
+
+    cmd="oc adm upgrade --to-image=${TARGET} --allow-explicit-upgrade"
+    if [[ "$SOURCE_VERSION" != *"nightly"* ]] && [[ "$TARGET_VERSION" != *"nightly"* ]]; then
+        if [[ "$ALLOW_NOT_RECOMMENDED" == 'true' ]]; then
+            run_command "$cmd --allow-not-recommended"
```

Comment:
If there is not Recommended update, but ALLOW_NOT_RECOMMENDED is set to true, it will fail or not? For ci, we never know if there will be recommended update when setting a job. 

There is a Recommended condition in conditionalUpdates. Maybe we could check if that condition exist/match before upgrade func.

Reply:
from the help:
```
    --allow-not-recommended=false:
	Allows upgrade to a version when it is supported but not recommended
	for updates.
```
I think `--allow-not-recommended=false` will not affect a normal upgrade, I add this parameter is just for future testing that we want to check unsafe upgrade.

Reply:
> run_command "$cmd --allow-not-recommended"

Here means --allow-not-recommended=true, right? Could we check if it  affect a normal upgrade?

Reply:
We don't have profiles that enabled `ALLOW_NOT_RECOMMENDED`, but anyway I can check it locally.

Reply:
Here is the test without and with --allow-not-recommended
```
[jianl@localhost 415]$ oc adm upgrade --to-image quay.io/openshift-release-dev/ocp-release@sha256:0da6316466d60a3a4535d5fed3589feb0391989982fba59d47d4c729912d6363 --allow-explicit-upgrade
warning: The requested upgrade image is not one of the available updates. You have used --allow-explicit-upgrade for the update to proceed anyway
Requested update to release image quay.io/openshift-release-dev/ocp-release@sha256:0da6316466d60a3a4535d5fed3589feb0391989982fba59d47d4c729912d6363
[jianl@localhost 415]$ 
[jianl@localhost 415]$ 
[jianl@localhost 415]$ oc adm upgrade --clear
Cancelled requested upgrade to quay.io/openshift-release-dev/ocp-release@sha256:0da6316466d60a3a4535d5fed3589feb0391989982fba59d47d4c729912d6363
[jianl@localhost 415]$ 
[jianl@localhost 415]$ oc adm upgrade --to-image quay.io/openshift-release-dev/ocp-release@sha256:0da6316466d60a3a4535d5fed3589feb0391989982fba59d47d4c729912d6363 --allow-explicit-upgrade --allow-not-recommended
warning: The requested upgrade image is not one of the available updates. You have used --allow-explicit-upgrade for the update to proceed anyway
Requested update to release image quay.io/openshift-release-dev/ocp-release@sha256:0da6316466d60a3a4535d5fed3589feb0391989982fba59d47d4c729912d6363
[jianl@localhost 415]$ 

```

Reply:
```
[jianl@localhost 415]$ oc get clusterversion
NAME      VERSION                              AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.15.0-0.nightly-2024-03-06-142724   True        True          61s     Working towards 4.15.0: 108 of 874 done (12% complete), waiting on etcd, kube-apiserver
[jianl@localhost 415]$ 
```

Reply:
I think we should add a separate step in `openshift-upgrade-qe-sanity` as the 1st step, once the target payload is not recommended, we can exit immediately to save the running for the following step in the chain.

Reply:
from the test result, it looks like a no harm var when there is not recommended update, so i'm ok to leave ALLOW_NOT_RECOMMENDED there.

Reply:
> once the target payload is not recommended, we can exit immediately to save the running for the following step in the chain.

+1 for this way, i also suggest to check it before upgrade with the recommended condition in my 1st comment.

Reply:
Agree with Jianlin, the intention is to exit as early as possible when we find the target image is a not recommended update.

Reply:
I think we need a call to discuss it

Reply:
Sure.